core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java

```diff
@@ -290,7 +292,9 @@ private void checkForRedundantAddedPartitions(PartitionField field) {
     ImmutableMap.Builder<Pair<Integer, String>, PartitionField> builder = ImmutableMap.builder();
     List<PartitionField> fields = spec.fields();
     for (PartitionField field : fields) {
-      builder.put(Pair.of(field.sourceId(), field.transform().toString()), field);
+      if (!field.transform().isVoid()) {
```

Comment:
Just a thought: would it make sense to just do `!(field.transform() instanceof VoidTransform)` here? Then you wouldn't need the extra interface method

Reply:
yes, in this case, they are same. i add the extra interface method because void transform is a particular transform, this method maybe used in other place later.  if we don't want to add extra interface method, i can change it to `!(field.transform() instanceof VoidTransform)`.

Reply:
Is this needed?

The map created by this method is only used in `addField` and `removeField`. My guess is that your intent is to avoid blocking multiple void transforms with the same source id because of multiple calls to remove. But the `apply` method works with the builder directly and so this map doesn't need to change if I'm reading the existing code correctly. What is the case where this causes a failure?

Reply:
ImmutableMap.Builder disallow multiple entries with same key, and multiple void transforms with the same source id generate same key here. So this builder will throw an exception with msg such as "Multiple entries with same key:  (1, void)=1001: create_time_month: void(1) and (1, void)=1000: create_time_year: void(1)".

Reply:
In that case, I think we should use `Maps.newHashMap` instead of `ImmutableMap.Builder`. That way we don't need to change the `Transform` API to add `isVoid`.

Reply:
Use Maps.newHashMap instead of ImmutableMap.Builder in newest commit. Please take a look again.